File: CodesBayesianAnalysis/normPostGibbs.R
# This program uses the Gibbs sampler for simulating from the posterior
# distribution of a normal model with unknown mean and variance. The mean
# and variance are given normal and Inv-chi2 priors. 

par(ask = TRUE);
normPostGibbs = function() {

	nIterations = 200;

	mu0 = 65;
	tau02 = 100^2;

	nu0 = 1;
	sigma02 = 1;

	yObs = c(72, 75, 70);
	n = length(yObs);

	mu = 50;
	sigma2 = 1;

	par(mfcol = c(1, 2));
	plot(mu, mfg = c(1, 1), xlab = "iteration", ylab =  "mu", type = 'l');
	plot(sqrt(sigma2), mfg = c(1, 2), xlab = "iteration", ylab =  "sigma", type = 'l');

	for(i in 2:nIterations) {
	# The conditional distribution of mu given sigma is normal(mu_n,
	# sigma_n) when we use a normal(mu0, tau02) prior for mu.
	mu_n = (mu0/tau02 + sum(yObs)/sigma2[i-1]) / (1/tau02 + n/sigma2[i-1]);
	sigma_n = sqrt(1/(1/tau02 + n/sigma2[i-1]));
	mu[i] = rnorm(1, mean = mu_n, sd = sigma_n);

	# The conditional distribution of variance sigma2 is Inv-chi2(nu_n, sigma02_n)
	# We can first sample z form chi2(nu_n) distribution then use nu_n*sigma02_n/z
	# as a sample from the scaled Inv-Chi2 distribution.  
	nu_n = nu0+n;
	nu = (sum( (yObs - mu[i])^2 ))/n;
	sigma02_n = (nu0*sigma02+n*nu)/(nu0+n);
	z = rchisq(1, nu_n);
	sigma2[i] = nu_n*sigma02_n/z;

		par(mfcol = c(1, 2));
		plot(mu, mfg = c(1, 1), xlab = "iteration", ylab =  "mu", type = 'l');
		plot(sqrt(sigma2), mfg = c(1, 2), xlab = "iteration", ylab =  "sigma", type = 'l');

	}
	return(list(mu = mu, sigma2 = sigma2));
}

param = normPostGibbs();

File: CodesBayesianAnalysis/normPostGibbs.m
% This program uses the Gibbs sampler for simulating from the posterior
% distribution of a normal model with unknown mean and variance. The mean
% and percision (i.e., 1/variance) are given normal and Gamma priors. 

function [mu, sigma2] = normPostGibbs
global yObs n mu0 sigma02 alpha beta;

nIterations = 200;

mu0 = 65;
sigma02 = 100^2;

alpha = 1;
beta = 1;

yObs = [72, 75, 70]';
[n, d] = size(yObs);

mu(1) = 50;
sigma2(1) = 1;

% This is to create a dynamic plot for MCMC sampling
h1 = subplot(2, 1, 1);
plot(mu);

xlabel(h1, '\iteration', 'FontSize', 18); ylabel(h1, '\mu', 'FontSize', 18);
axis tight
grid off

h2 = subplot(2, 1, 1);
plot(1, sqrt(sigma2));

xlabel(h2, '\iteration', 'FontSize', 18); ylabel(h2, '\sigma', 'FontSize', 18);
axis tight
grid off

for i = 2:nIterations

    % The conditional distribution of mu given sigma is normal(mu_n,
    % sigma_n) when we use a normal(mu0, sigma02) prior for mu.
    mu_n = (mu0/sigma02 + sum(yObs)/sigma2(i-1)) / (1/sigma02 + n/sigma2(i-1));
    sigma_n = sqrt(1/(1/sigma02 + n/sigma2(i-1)));
    mu(i) = normrnd(mu_n, sigma_n);
    
    % The conditional distribution of precision (i.e., 1/sigma2) is Gamma 
    % when we use a Gamma(alpha, beta) prior. Note that if you are using R,
    % you have to inverse beta, and also inverse the second parameter of the
    % posterior distribution.
    precision = gamrnd(alpha + n/2, 1/(1/beta + sum( (yObs - mu(i)).^2 )/2));
    sigma2(i) = 1/precision;
    
    % Ploting the samples
    h1 = subplot(2, 1, 1);    
    plot(mu)
    xlabel(h1, 'iteration', 'FontSize', 18); ylabel(h1, '\mu', 'FontSize', 18);
   
    h2 = subplot(2, 1, 2);    
    plot(sqrt(sigma2))

    xlabel(h2, 'iteration', 'FontSize', 18); ylabel(h2, '\sigma', 'FontSize', 18);

    drawnow
        
end



